File: Data Structures & Algorithms/valid-tree/submission-1.java
class Solution {
    public boolean validTree(int n, int[][] edges) {
        
        List<Integer> g[] =new ArrayList[n];

        for(int i=0;i<n;i++)
        {
            g[i] = new ArrayList<>();
        }

        for(int[] edg: edges)
        {
            g[edg[0]].add(edg[1]);
            g[edg[1]].add(edg[0]);
        }
        Set<Integer> visited=new HashSet<>();
        
        if(!dfs(0,-1,g,visited))
                return false;
        return visited.size()==n;
    }

    private boolean dfs(int u, int parent, List<Integer>[] g, Set<Integer> visited)
    {
        if(visited.contains(u))
            return false;

        visited.add(u);

        for(int i :g[u])
        {
            if(i==parent)
                continue;
            if(!dfs(i,u,g,visited))
                return false;
        }
        return true;
    }
}

File: Data Structures & Algorithms/redundant-connection/submission-1.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:

        n = len(edges)

        graph = [[] for i in range(n+1)]

        
        
        def dfs(u, par):

            if visited[u] == True:
                return True
            
            visited[u] = True
            for  v in graph[u]:
                if v == par:
                    continue
                
                if dfs(v,u):
                    return True
            return False
        
        for u,v in edges:
            graph[u].append(v)
            graph[v].append(u)

            visited = [False] * (n+1)

            if dfs(u,-1):
                return [u,v]
        
        return []
        